src/cost_estimates/pessimistic.rs

```diff
@@ -0,0 +1,304 @@
+                        tx.commit()?;
+                        Ok(db)
+                    } else {
+                        Err(e)
+                    }
+                } else {
+                    Err(e)
+                }
+            })?;
+        Ok(PessimisticEstimator { db, log_error })
+    }
+
+    fn instantiate_db(tx: &SqliteTransaction) -> Result<(), SqliteError> {
+        tx.execute(CREATE_TABLE, rusqlite::NO_PARAMS)?;
+        Ok(())
+    }
+
+    fn get_estimate_key(tx: &TransactionPayload, field: &CostField) -> String {
+        let tx_descriptor = match tx {
+            TransactionPayload::TokenTransfer(..) => "stx-transfer".to_string(),
+            TransactionPayload::ContractCall(cc) => {
+                format!("cc:{}.{}", cc.contract_name, cc.function_name)
+            }
+            TransactionPayload::SmartContract(_sc) => "contract-publish".to_string(),
+            TransactionPayload::PoisonMicroblock(_, _) => "poison-ublock".to_string(),
+            TransactionPayload::Coinbase(_) => "coinbase".to_string(),
+        };
+
+        format!("{}:{}", &tx_descriptor, field)
+    }
+}
+
+impl From<SqliteError> for EstimatorError {
+    fn from(e: SqliteError) -> Self {
+        EstimatorError::SqliteError(e)
+    }
+}
+
+impl CostEstimator for PessimisticEstimator {
+    fn notify_event(
+        &mut self,
+        tx: &TransactionPayload,
+        actual_cost: &ExecutionCost,
+    ) -> Result<(), EstimatorError> {
+        if self.log_error {
+            // only log the estimate error if an estimate could be constructed
+            if let Ok(estimated_cost) = self.estimate_cost(tx) {
+                let estimated_scalar =
+                    estimated_cost.proportion_dot_product(&BLOCK_LIMIT_MAINNET, 1_000);
+                let actual_scalar = actual_cost.proportion_dot_product(&BLOCK_LIMIT_MAINNET, 1_000);
+                info!("PessimisticEstimator received event";
+                      "key" => %PessimisticEstimator::get_estimate_key(tx, &CostField::RuntimeCost),
+                      "estimate" => estimated_scalar,
+                      "actual" => actual_scalar,
+                      "estimate_err" => (estimated_scalar as i64 - actual_scalar as i64),
+                      "estimate_err_pct" => (estimated_scalar as i64 - actual_scalar as i64)/(cmp::max(1, actual_scalar as i64)),);
+                for field in CostField::ALL.iter() {
+                    info!("New data event received";
+                          "key" => %PessimisticEstimator::get_estimate_key(tx, field),
+                          "value" => field.select_key(actual_cost));
+                }
+            }
+        }
+
+        let sql_tx = tx_begin_immediate_sqlite(&mut self.db)?;
+        for field in CostField::ALL.iter() {
+            let key = PessimisticEstimator::get_estimate_key(tx, field);
+            let field_cost = field.select_key(actual_cost);
+            let mut current_sample = Samples::get_sqlite(&sql_tx, &key);
```

Comment:
It doesn't appear to be particularly costly -- simple gets / inserts in SQLite are much faster than most other kinds of operations performed while processing blocks, and in spinning up nodes that use this feature branch, I haven't noticed any performance issues related to this path. If this did lead to a slow down in node spinup, though, I think it would be useful to benchmark.